### ISSUE TITLE
Isolation of query planning

### ISSUE THREAD

**Issue by teskje** [MaterializeInc/materialize#16217]
We regularly find bugs around query planning (https://github.com/MaterializeInc/materialize/issues/15570, https://github.com/MaterializeInc/materialize/issues/16128, https://github.com/MaterializeInc/materialize/issues/16195, ...) that, when triggered, produce a panic which causes the environmentd process to crash with a panic or a stack overflow. We also find performance issues where the optimizer just spins for a long time without yielding, effectively blocking the whole coordinator for that time (#16204). Of course we want to fix such bugs, but it is unrealistic to expect that we somehow find and remove all of them.

We should think about ways to make Materialize more robust to errors during query planning (mostly optimization, AFAICT). Ideally, if a user runs a SQL statement that triggers a bug in our code, the resulting crash or performance degredation should effect only that user's session, not all users of the same environment. We want some kind of isolation between users.

In its simplest form, that isolation could consist of `catch_unwind`s at strategic places (e.g. around optimizer calls), that ensure that crashes in some parts of the code can be handled gracefully, e.g., by returning an error to the user. This is not trivial to implement in our current code, because a) we abort on panics, so there is no opportunity to catch unwinds, and b) it's not clear where `catch_unwind`s should be added to ensure that panics are caught in most cases, while at the same time ensuring that we don't leave behind inconsistent state anywhere. This solution is also not a complete one, since it doesn't help with segfaults caused by stack overflows, OOM conditions, or the optimizer hogging the executor for several minutes.

A complete solution could be to move parts of the query planning into a separate subprocess. That should work at least for SQL parsing and optimization, which both should not need to mutate the coordinator state. Crashes of subprocesses wouldn't affect the parent (environmentd), and neither would excessive memory or CPU usage (provided we manage to set per-process limits).

**Comment by mjibson** [comment]
I've done some initial work on being able to move the work in sequence_peek (which also calls the optimizer) into a separate thread. Yesterday I found out that the one blocker I had is fixed, so I can do more of it. Some details in https://github.com/MaterializeInc/materialize/pull/15432#issuecomment-1304980913

**Comment by benesch** [comment]
> A complete solution could be to move parts of the query planning into a separate subprocess. That should work at least for SQL parsing and optimization, which both should not need to mutate the coordinator state. Crashes of subprocesses wouldn't affect the parent (environmentd), and neither would excessive memory or CPU usage (provided we manage to set per-process limits).

I would advocate for separate tasks/threads rather than subprocesses. At least, the most compelling reason to switch to subprocesses IMO is to facilitate horizontal scaling of `environmentd`. I worry that if we switch to subprocesses earlier, to facilitate fault isolation, we'll cut the boundaries differently, and in a way that inhibits delivering on horizontal scaling later.

> This is not trivial to implement in our current code, because a) we abort on panics, so there is no opportunity to catch unwinds, and b) it's not clear where catch_unwinds should be added to ensure that panics are caught in most cases, while at the same time ensuring that we don't leave behind inconsistent state anywhere. This solution is also not a complete one, since it doesn't help with segfaults caused by stack overflows, OOM conditions, or the optimizer hogging the executor for several minutes.

(a) I am definitely in favor of fixing. We should absolutely consider alternative ways of handling panics. The only thing I care about is that panics in threads/tasks result in taking down the process _unless_ they are explicitly handled.

(b) Is "around every call to the transform crate" not a good heuristic? My model is that the optimizer is a pure function and so it's always safe to catch and discard panics that originate in the optimizer (as long as you discard the buggy query, too), since the optimizer can't possibly have left any of the rest of the system in an inconsistent state.

**Comment by teskje** [comment]
> I would advocate for separate tasks/threads rather than subprocesses.

Scaling of environmentd is something I had not considered, but your concern makes sense. Using threads and catching their panics solves some of the issues we currently have (crashes caused by panics and the optimizer not yielding to the executor), but not all of them (crashes caused by stack overflows (?), the process getting OOM-killed). Still, they would be a huge improvement over the situation we have now.

> (a) I am definitely in favor of fixing. We should absolutely consider alternative ways of handling panics. The only thing I care about is that panics in threads/tasks result in taking down the process unless they are explicitly handled.

I've opened a separate ticket for this: https://github.com/MaterializeInc/materialize/issues/16236. Seems like a nice skunkworks project, or even something for the "stability" roadmap.

> (b) Is "around every call to the transform crate" not a good heuristic?

Maybe! I think we can start with that and add more `catch_unwinds` as we encounter environmentd crashes over time. My main concern is that we don't litter the adapter code with `catch_unwinds` everywhere. Ideally, the optimizer methods would handle panics themselves and convert them all to `Result::Err`.

**Comment by benesch** [comment]
> but not all of them (crashes caused by stack overflows (?), the process getting OOM-killed). Still, they would be a huge improvement over the situation we have now.

Stack overflows don't worry me too much! The checked recursion seems to work well in practice. And if we're seeing a lot a stack overflows, it is in theory safe to catch a stack overflow in a thread and restart the thread without restarting the whole process, though I'm sure Rust makes it difficult.

Running out of memory is tricky for sure though. If the optimizer becomes very memory hungry, to the point that we need to limit the amount of memory available per query/connection/something, yeah, I think we'll need to consider process isolation. But I think that'll actually fall under the "environmentd scalability" heading more than the "environmentd stability" heading. We're not currently seeing bugs that cause pathological memory usage in the optimizer that crashes environmentd due to OOM, are we?

> Ideally, the optimizer methods would handle panics themselves and convert them all to `Result::Err`.

Agreed!

**Comment by benesch** [comment]
With [#16263](https://github.com/MaterializeInc/materialize/pull/16263) merged, I believe we're unblocked here to take the next step of a few well-placed `mz_ore::panic::catch_unwind` calls in the optimizer.

**Comment by teskje** [comment]
I created an issue to track catching optimizer panics here: https://github.com/MaterializeInc/materialize/issues/16324

**Comment by philip-stoev** [comment]
To reproduce the fact that query optimization currently happens on the critical path:

```
materialize=> select mz_internal.mz_sleep(90000);
```
Mz is now borked and no future queries can run. Stack trace currently is:
```
...
#26 0x000055f522f84d1e in mz_transform::Optimizer::optimize ()
#27 0x000055f5226b7620 in mz_adapter::coord::sequencer::<impl mz_adapter::coord::Coordinator>::plan_peek ()
#28 0x000055f5219d3cde in mz_adapter::coord::sequencer::<impl mz_adapter::coord::Coordinator>::sequence_peek_finish::{{closure}}::{{closure}} ()
#29 0x000055f5219d355a in mz_adapter::coord::sequencer::<impl mz_adapter::coord::Coordinator>::sequence_peek_finish::{{closure}} ()
#30 0x000055f5219ca39b in mz_adapter::coord::sequencer::<impl mz_adapter::coord::Coordinator>::sequence_peek_begin::{{closure}}::{{closure}} ()
#31 0x000055f5219c8b50 in mz_adapter::coord::sequencer::<impl mz_adapter::coord::Coordinator>::sequence_peek_begin::{{closure}} ()
#32 0x000055f5219b7851 in mz_adapter::coord::sequencer::<impl mz_adapter::coord::Coordinator>::sequence_plan::{{closure}}::{{closure}} ()
#33 0x000055f5219ae66d in mz_adapter::coord::sequencer::<impl mz_adapter::coord::Coordinator>::sequence_plan::{{closure}} ()
#34 0x000055f52183ed5c in mz_adapter::coord::command_handler::<impl mz_adapter::coord::Coordinator>::handle_execute_inner::{{closure}}::{{closure}} ()
#35 0x000055f52183d71e in mz_adapter::coord::command_handler::<impl mz_adapter::coord::Coordinator>::handle_execute_inner::{{closure}} ()
#36 0x000055f52183a979 in mz_adapter::coord::command_handler::<impl mz_adapter::coord::Coordinator>::handle_execute::{{closure}}::{{closure}} ()
#37 0x000055f521839b44 in mz_adapter::coord::command_handler::<impl mz_adapter::coord::Coordinator>::handle_execute::{{closure}} ()
#38 0x000055f521818d78 in <tracing::instrument::Instrumented<T> as core::future::future::Future>::poll ()
#39 0x000055f521838634 in mz_adapter::coord::command_handler::<impl mz_adapter::coord::Coordinator>::handle_command::{{closure}} ()
#40 0x000055f520df11ff in mz_adapter::coord::message_handler::<impl mz_adapter::coord::Coordinator>::message_command::{{closure}}::{{closure}} ()
#41 0x000055f520df02d7 in mz_adapter::coord::message_handler::<impl mz_adapter::coord::Coordinator>::message_command::{{closure}} ()
#42 0x000055f520def75d in mz_adapter::coord::message_handler::<impl mz_adapter::coord::Coordinator>::handle_message::{{closure}} ()
#43 0x000055f5208c52d5 in mz_adapter::coord::Coordinator::serve::{{closure}} ()
...
```

**Comment by mjibson** [comment]
#18994 is merged so now peeks are optimized in a separate task.

**Comment by benesch** [comment]
May be known, but views are still optimized on the main thread! So `CREATE VIEW v AS SELECT mz_internal.mz_sleep(100000)` can still perma-wedge the coordinator.

**Comment by aalexandrov** [comment]
Summarizing some important points form a related Slack discussion:

We want to ensure that:

1. We cancel cleanly after X amount of seconds.
1. We cancel cleanly if the client session running the DDL receives a SIGTERM / CTRL + C while handling the sequence_create_~ call.

@mjibson:

> The two listed points there will use roughly the same mechanism (cancellation). `SIGTERM` or `ctrl + c` are things only `psql` cares about, after which it triggers a cancellation request. That request works by `psql` opening a whole new connection and sending a special startup message asking to cancel its other connection (we also support `select pg_cancel_backend(x) ` to do this, but that still requires another connection). That connection sends a cancellation request to the coordinator which then triggers a few things. Each place in the code that we would like to support cancellation has to be explicitly taught how to listen for this while doing its work, so we'd need to teach this new off thread stuff how to do that. The optimizer itself never checks (and it's a large amount of work to teach it to), so this means we'd probably need to start yet another task that listens for the timeout from your point #1, listens for cancel from point #2, and also is waiting for the optimizer task to return, then it would return up whatever happens first.

---

@ParkMyCar:

> Once you call `spawn_blocking(...)` the provided closure will get executed, regardless of whether or not the returned `JoinHandle` gets polled.

---

@benesch:
> The way to make the optimizer support cancelation is to plumb a cancelation token (is_canceled: AtomicBool) through the entire optimizer, and make sure you check it periodically while doing expensive things. Simple thing would be to check it between every call to transform, but you might also want to check it in hot loops/traversals within a given transform. That’s harder though.

> I feel like we _could_ use more coord scaffolding around off-thread work.

> Some combo of helper methods/traits or something that makes the pattern of “on thread work to generate a description of work to handle off thread, plus a handler function to handle that work off thread then then produces an internal command, then an internal command handler that does the remaining work on thread, … rinse and repeat” less painful to type in.
